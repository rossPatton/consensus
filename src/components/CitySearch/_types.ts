export type tEventTypes = React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>;
type tUpdateState = (key: any, value: string | number | object | boolean) => void;

// these are inherited by the form
type tProps = {
  city: string,
  cityId: number | null,
  geo: tGeo,
  region: string,
  regionId: number | null,
  session: tSession,
};

export type tContainerProps = tProps & {
  citiesThunk: tThunk<tCity[]>,
  getCitiesDispatch: (query?: {region: string}) => tThunkPayload<tCity[]>,
  // inherited. updates form state
  updateState: tUpdateState,
};

export type tState = {showRegionField: boolean};
export type tKeyUnion = 'city' | 'cityId' | 'region';

export type tComponentProps = tState & tProps & {
  cities: tCity[],
  // local update state. just updates the city and region inputs
  updateState: (state: Partial<tProps>) => void,
  toggleRegionField: () => void,
};

export type tStore = {
  cities: tThunk<tCity[]>,
  geo: tThunk<tGeo>,
  session: tThunk<tSession>,
};
