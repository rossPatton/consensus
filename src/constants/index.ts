export const categories: ts.categoryObj[] = [
  {display: 'Community', slug: 'community'},
  {display: 'Cooperative', slug: 'cooperative' },
  {display: 'Union', slug: 'union' },
  {display: 'Political', slug: 'political' },
];

export const categoryMap: ts.categoryMap = {
  'community': 'Community',
  'cooperative': 'Cooperative',
  'union': 'Union',
  'political': 'Political',
};

export const roles: ts.role[] = ['member', 'facilitator'];

export const spacesUrl = 'https://consensus.nyc3.digitaloceanspaces.com';
