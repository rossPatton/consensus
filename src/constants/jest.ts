export const testGroup1 = {
  avatar: '',
  category: 'Union' as ts.category,
  city: 'New York',
  cityId: 1,
  country: 'United States',
  countryId: 1,
  description: '',
  deletionDeadline: '',
  email: 'testgroup1@test.com',
  id: 1,
  memberName: 'member',
  modName: 'facilitator',
  name: 'Tech Workers Coalition Chapter',
  region: 'New York',
  regionId: 1,
  handle: 'tech-workers-coalition-chapter',
  sessionType: 'group' as ts.sessionType,
  type: 'private' as ts.privacyEnum,
};

export const testMeeting1 = {
  attendees: 1,
  host: '',
  publicRSVPS: [{}] as ts.user[],
  rsvp: { userId: 1 },
  id: 12,
  category: 'Political' as ts.category,
  groupId: 1,
  cityId: 16624,
  groupName: 'Tech Workers Coalition NYC',
  isOnline: false,
  isPrivate: false,
  description: 'Asperiores et perspiciatis fugit sint. Nihil dolore autem enim numquam dolores aliquam sed repudiandae. Eius dolorem nostrum quo aut culpa incidunt. Aut eum adipisci. Amet amet suscipit et. Voluptas animi distinctio adipisci. Nam voluptas similique quam esse magni quidem id. In aut nulla inventore facilis dicta ducimus molestias saepe. Autem voluptatibus mollitia quos ullam quidem ex consequuntur quia. Adipisci quod est incidunt quas repellendus rem voluptatem. Cumque iusto corporis velit nam nostrum commodi vel possimus velit. Quibusdam ea cumque dignissimos cumque magni voluptas ipsa natus sunt. Est qui officiis atque dolorem voluptatibus aut explicabo.',
  location: '717 Borer Oval',
  locationLink: 'http://sam.info',
  locationType: 'online',
  title: 'aut occaecati iusto nostrum ut',
  slug: 'aut-occaecati-iusto-nostrum-ut',
  date: '2020-05-02T10:47:34.393Z',
  endDate: '2021-02-12T02:01:06.513Z',
  isDraft: false,
  duration: 2,
  tag: 'Protest' as ts.meetingTypes,
  img: '',
  time: '19:00',
};

export const testMeeting2 = {
  attendees: 1,
  category: 'Cooperative' as ts.category,
  cityId: 16624,
  date: '2020-05-02T10:47:34.393Z',
  description: 'Asperiores et perspiciatis fugit sint. Nihil dolore autem enim numquam dolores aliquam sed repudiandae. Eius dolorem nostrum quo aut culpa incidunt. Aut eum adipisci. Amet amet suscipit et. Voluptas animi distinctio adipisci. Nam voluptas similique quam esse magni quidem id. In aut nulla inventore facilis dicta ducimus molestias saepe. Autem voluptatibus mollitia quos ullam quidem ex consequuntur quia. Adipisci quod est incidunt quas repellendus rem voluptatem. Cumque iusto corporis velit nam nostrum commodi vel possimus velit. Quibusdam ea cumque dignissimos cumque magni voluptas ipsa natus sunt. Est qui officiis atque dolorem voluptatibus aut explicabo.',
  duration: 2,
  endDate: '2021-02-12T02:01:06.513Z',
  groupId: 1,
  groupName: 'Test Group',
  host: '',
  id: 12,
  img: '',
  isDraft: false,
  isOnline: false,
  isPrivate: false,
  location: '717 Borer Oval',
  locationLink: 'http://sam.info',
  locationType: 'online',
  publicRSVPS: [{}] as ts.user[],
  rsvp: { userId: 1 },
  slug: 'aut-occaecati-iusto-nostrum-ut',
  tag: 'Protest' as ts.meetingTypes,
  time: '19:00',
  title: 'aut occaecati iusto nostrum ut',
};

export const testGeo = {
  city: 'New York',
  countryCode: 'us',
  handle: 'new-york',
  region: 'New York',
  regionCode: 'ny',
  postcode: 10002,
};

export const testNYC = {
  id: 1,
  name: 'New York',
  country: 'United States',
  countryId: 1,
  region: 'New York',
  regionId: 37,
  regionCode: 'ny',
};
