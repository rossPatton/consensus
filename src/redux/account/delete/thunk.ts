import { path } from '../_constants';
import { api } from '../../../utils';
import { failure, success } from './actions';

export const deleteAccount = () => {
  return async function (dispatch: Function) {
    return api({
      credentials: true,
      dispatch,
      failure,
      method: 'DELETE',
      path,
      success,
    });
  };
};
