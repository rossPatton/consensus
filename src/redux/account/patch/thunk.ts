import { api } from '@app/utils';

import { path } from '../_constants';
import { failure, success } from './actions';

export const patchAccount = (query: tAccountQuery) => {
  return async function (dispatch: Function) {
    return api({
      credentials: true,
      dispatch,
      failure,
      method: 'PATCH',
      query,
      path,
      success,
    });
  };
};
