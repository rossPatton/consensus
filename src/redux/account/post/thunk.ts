import { api } from '~app/utils';

import { path } from '../_constants';
import { failure, init, success } from './actions';

export const postAccount = (query: ts.accountQuery) => {
  return async function (dispatch: Function) {
    return api({
      credentials: true,
      dispatch,
      failure,
      init,
      method: 'POST',
      query,
      path,
      success,
    });
  };
};
