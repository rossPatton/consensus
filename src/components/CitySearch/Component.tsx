import _ from 'lodash';
import React, {memo} from 'react';

import stateCodeMap from '../../json/usa/stateCodeMap.json';
import {tComponentProps} from './_types';

export const CitySearchComponent = memo((props: tComponentProps) => {
  const onCityChange = (ev: React.ChangeEvent<HTMLSelectElement>) => {
    const city = _.find(props.cities, c => c.name === ev.currentTarget.value);
    if (city) {
      props.updateState({
        city: city.name,
        cityId: city.id,
        region: city.region,
        regionId: city.regionId,
      });
    }
  };

  const onRegionChange = (ev: React.ChangeEvent<HTMLSelectElement>) => {
    const region = ev.currentTarget.value;
    if (region !== props.region) {
      props.updateState({
        city: '',
        cityId: 0,
        region,
        regionId: 0,
      });
    }
  };

  return (
    <>
      {props.showRegionField && (
        <>
          <h2 className="fs5 mB1 lh1">
            Pick a different state
          </h2>
          <label htmlFor="stateSelect">
            <select
              name="type"
              id="stateSelect"
              className="mB4 row bgWhite"
              value={props.region}
              onBlur={onRegionChange}
              onChange={onRegionChange}>
              {Object.keys(stateCodeMap).map(state => (
                <option key={state} value={state}>
                  {state}
                </option>
              ))}
            </select>
          </label>
        </>
      )}
      <h2 className="fs5 mB1 lh1">
        City in <span className="dInBl mR2">
          {(!props.showRegionField && props.geo.region) && props.geo.region}
          {(props.showRegionField && props.region) && props.region}
        </span>
        {!props.showRegionField && (
          <button
            type="button"
            onClick={props.toggleRegionField}>
            Not the right state?
          </button>
        )}
      </h2>
      {props.label && (
        <p className="fs5 copyBlack mB1">
          {props.label}
        </p>
      )}
      <div className="mB4">
        {props.cityId === 0 && (
          <label htmlFor="citySelect">
            <select
              name="type"
              id="citySelect"
              className="row bgWhite"
              value={props.city}
              onBlur={onCityChange}
              onChange={onCityChange}>
              {props.cities.map((city: tCity) => (
                <option key={city.id} value={city.name}>
                  {city.name}, {city.region}
                </option>
              ))}
            </select>
          </label>
        )}
        {props.cityId > 0 && (
          <div className="brdA1 p3 black br8 dInBl">
            <b>{props.city}</b>, <span className="dInBl mR3">{props.region}</span>
            <button
              type="button"
              onClick={() => {
                props.updateState({
                  city: '',
                  cityId: 0,
                });
              }}>
              Not the right city?
            </button>
          </div>
        )}
      </div>
    </>
  );
});
