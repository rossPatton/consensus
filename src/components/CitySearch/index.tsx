import _ from 'lodash';
import React, {PureComponent} from 'react';
import {connect} from 'react-redux';

import {GenericLoader} from '~app/containers';
import {getCities} from '~app/redux';

import {tContainerProps, tStore} from './_types';
import {CitySearchComponent} from './Component';

// use within an existing form, tie to it's updateState method
class CitySearchContainer extends PureComponent<tContainerProps> {
  static defaultProps = {
    city: 'New York',
    region: 'New York',
    showResetButton: true,
  };

  constructor(props: tContainerProps) {
    super(props);
    let region = 'New York';
    const sessionRegion = _.get(props, 'session.profile.region', null);
    const geoRegion = _.get(props, 'geo.region', null);

    if (typeof sessionRegion === 'string' && sessionRegion !== '') {
      region = sessionRegion;
    } else if (typeof geoRegion === 'string' && geoRegion !== '') {
      region = geoRegion;
    }

    props.getCitiesDispatch({ region });
  }

  updateState = async (state: {[key: string]: unknown}) => {
    if (state.region) {
      await this.props.getCitiesDispatch({region: state.region as string});
    }

    // set local state for the city input, but then pass it up to the form
    this.props.updateState(null, state);
  }

  render() {
    return (
      <GenericLoader
        isLoading={this.props.citiesThunk.isLoading}
        render={() => (
          <CitySearchComponent
            {...this.props}
            cities={this.props.citiesThunk.data}
            geo={this.props.geo}
            session={this.props.session}
            updateState={this.updateState}
          />
        )}
      />
    );
  }
}

const mapStateToProps = (store: tStore) => ({
  citiesThunk: store.cities,
  geo: store.geo.data,
  session: store.session.data,
});

const mapDispatchToProps = (dispatch: Function) => ({
  getCitiesDispatch: (query: {region: string}) => dispatch(getCities(query)),
});

const CitySearch = connect(
  mapStateToProps,
  mapDispatchToProps,
)(CitySearchContainer);

export default CitySearch;
