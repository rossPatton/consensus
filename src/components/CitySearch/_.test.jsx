import React from 'react';
import { Provider } from 'react-redux';
import render from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import {testGeo, testNYC} from '~app/constants/jest';

import CitySearch from '.';

const mockStore = configureStore([thunk]);
const defaultStore = {
  cities: {
    error: null,
    isLoading: false,
    data: [testNYC],
  },
  geo: {
    error: null,
    isLoading: false,
    data: testGeo,
  },
  session: {
    error: null,
    isLoading: false,
    data: {
      isAuthenticated: false,
      profile: {
        region: 'New York',
      },
    },
  },
};

describe('components/CitySearch', () => {
  let store = mockStore({});

  beforeEach(() => {
    store = mockStore(defaultStore);
  });

  it('renders without crashing', () => {
    const component = render.create((
      <Provider store={store}>
        <CitySearch
          city="New York"
          cityId={16625}
          getCitiesDispatch={jest.fn()}
          region="New York"
          regionId={37}
        />
      </Provider>
    ));
    expect(component).toMatchSnapshot();
  });

  it('renders with custom label', () => {
    const component = render.create((
      <Provider store={store}>
        <CitySearch
          city="New York"
          cityId={16625}
          getCitiesDispatch={jest.fn()}
          label="Testing custom label"
          region="New York"
          regionId={37}
        />
      </Provider>
    ));
    expect(component).toMatchSnapshot();
  });

  it('renders with remove button', () => {
    const component = render.create((
      <Provider store={store}>
        <CitySearch
          city="New York"
          cityId={16625}
          getCitiesDispatch={jest.fn()}
          region="New York"
          regionId={37}
          showRemoveButton
        />
      </Provider>
    ));
    expect(component).toMatchSnapshot();
  });

  it('renders with blank session region', () => {
    const customStore = {
      ...store,
      session: {
        error: null,
        isLoading: false,
        data: {
          isAuthenticated: false,
          profile: {
            region: '',
          },
        },
      },
    };
    const component = render.create((
      <Provider store={store}>
        <CitySearch
          city="New York"
          cityId={16625}
          getCitiesDispatch={jest.fn()}
          region="New York"
          regionId={37}
          showRemoveButton
        />
      </Provider>
    ));
    expect(component).toMatchSnapshot();
  });

  it('renders with null session region', () => {
    const customStore = {
      ...store,
      session: {
        error: null,
        isLoading: false,
        data: {
          isAuthenticated: false,
          profile: {
            region: null,
          },
        },
      },
    };
    const component = render.create((
      <Provider store={store}>
        <CitySearch
          city="New York"
          cityId={16625}
          getCitiesDispatch={jest.fn()}
          region="New York"
          regionId={37}
          showRemoveButton
        />
      </Provider>
    ));
    expect(component).toMatchSnapshot();
  });
});
