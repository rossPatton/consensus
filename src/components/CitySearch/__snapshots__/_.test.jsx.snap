// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CitySearch renders with blank session region 1`] = `null`;

exports[`components/CitySearch renders with custom label 1`] = `null`;

exports[`components/CitySearch renders with null session region 1`] = `null`;

exports[`components/CitySearch renders with remove button 1`] = `null`;

exports[`components/CitySearch renders without crashing 1`] = `null`;
